feat: implemented pokemon details page

jayveeannsoriano opened the pull request:
- added a synchronous action to dispose pokemon details page state
- created and added vm and connector for pokemon details page state and UI
- added barrel files to shorten import lines
- edited abilites model jsonkey
- created and added reusable widgets for pokemon details page
- added color_constants.dart in utils folder
- added a GestureDetector for pokemon_tile_card.widget
- added a private function for the GestureDetector to navigate to pokemon details page

**Video attachment details:** 

- tapping on a pokemon from the list will navigate the user to the respective pokemon's details

-  this will also show an error if the pokemon details page state is null

https://user-images.githubusercontent.com/111727585/228356911-8659ff0d-faa0-4dc2-ab97-f66ee81d5332.mp4

- flutter lint analyzer results and format
![image](https://user-images.githubusercontent.com/111727585/228357969-9843b7be-2c6d-4951-9bd0-4d9b25fb5b24.png)

Commit:
feat: implemented pokemon details page
- added a synchronous action to dispose pokemon details page state
- created and added vm and connector for pokemon details page state and UI
- added barrel files to shorten import lines
- edited abilites model jsonkey
- created and added reusable widgets for pokemon details page
- added color_constants.dart in utils folder
- added a GestureDetector for pokemon_tile_card.widget
- added a private function for the GestureDetector to navigate to pokemon details page

shawn-ffufm commented on lib/api/model/abilities.dart:
```diff
@@ -7,7 +7,7 @@ part 'abilities.g.dart';
 class Abilities with _$Abilities {
   factory Abilities({
-    @JsonKey(name: 'move') AbilityDetails? move,
+    @JsonKey(name: 'ability') AbilityDetails? ability,
```
can you confirm if this will really going to be null in some other pokemons? 🤔

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details.dart:
```diff
@@ -0,0 +1,3 @@
+export 'pokemon_details_connector.dart';
```
do you really need this wild card import? remember it's gonna force to import other classes as well even if for example the page class doesn't need to know anything about vm 🙂

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_connector.dart:
```diff
@@ -0,0 +1,29 @@
+    super.key,
+  });
+
+  final String pokemonName;
```
why not pass the whole pokemon object instead since these properties u indicated are all from pokemon model? 🙂

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_connector.dart:
```diff
@@ -0,0 +1,29 @@
+      onDispose: (store) => store.dispatchSync(ClearPokemonDetailsAction()),
+      builder: (context, vm) => PokemonDetailsPage(
+        pokemonDetails: vm.pokemonDetails,
+        pokemonName: pokemonName,
```
use the whole pokemon model instead, then place it above `pokemonDetails` property 🙂

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_page.dart:
```diff
@@ -0,0 +1,47 @@
+  });
+
+  final Async<PokemonDetails> pokemonDetails;
+  final String pokemonName;
```
use the whole pokemon model instead, then place it above `pokemonDetails` property 🙂

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_page.dart:
```diff
@@ -0,0 +1,47 @@
+      body: pokemonDetails.when(
+        (data) => Padding(
+          padding: const EdgeInsets.all(20.0),
+          child: SingleChildScrollView(
```
use ListView instead since the properties you've used here are dynamic value depend on what particular pokemon you're gonna see its details 🙂

shawn-ffufm commented on lib/features/pokemon_details/widgets/pokemon_abilities_list.dart:
```diff
@@ -0,0 +1,35 @@
+import 'package:pokedex_async_redux/utils/constants.dart';
+
+class PokemonAbilitiesList extends StatelessWidget {
+  const PokemonAbilitiesList({
```
maybe u can rename this to PokemonAbilitiesView, same for its dart file 🙂

shawn-ffufm commented on lib/features/pokemon_details/widgets/pokemon_details_overview.dart:
```diff
@@ -0,0 +1,34 @@
+import 'package:flutter/material.dart';
+
+class PokemonDetailsOverview extends StatelessWidget {
+  const PokemonDetailsOverview({
```
seems no need for this class anymore, just directly place these components in your `pokemon_details_page` 🙂

shawn-ffufm commented on lib/features/pokemon_details/widgets/pokemon_moves_list.dart:
```diff
@@ -0,0 +1,52 @@
+import 'package:flutter/material.dart';
+
+class PokemonMovesList extends StatelessWidget {
+  const PokemonMovesList({
```
maybe u can rename this to `PokemonMovesView`, same for its dart file 🙂

shawn-ffufm commented on lib/features/pokemon_details/widgets/pokemon_physical_stats.dart:
```diff
@@ -0,0 +1,69 @@
+import 'package:flutter/material.dart';
+
+class PokemonPhysicalStats extends StatelessWidget {
+  const PokemonPhysicalStats({
```
maybe u can rename this to `PokemonPhysicalStatsView`, same for its dart file 🙂

shawn-ffufm commented on lib/features/pokemon_details/widgets/pokemon_physical_stats.dart:
```diff
@@ -0,0 +1,69 @@
+  }
+}
+
+class PhysicalStatsColumn extends StatelessWidget {
```
kindly exclude this in a separate dart file and rename it `PhysicalStatsItem` 🙂

shawn-ffufm commented on lib/features/pokemon_details/widgets/pokemon_stats_list.dart:
```diff
@@ -0,0 +1,69 @@
+import 'package:flutter/material.dart';
+
+class PokemonStatsList extends StatelessWidget {
+  const PokemonStatsList({
```
maybe u can rename this to `PokemonStatsView`, same for its dart file 🙂

shawn-ffufm commented on lib/features/pokemon_details/widgets/widgets.dart:
```diff
@@ -0,0 +1,5 @@
+export 'pokemon_abilities_list.dart';
```
no need for this, as much as possible avoid this kind of wild cart import 🙈

shawn-ffufm commented on lib/state/state.dart:
```diff
@@ -0,0 +1,3 @@
+export 'action/actions.dart';
```
same here, see related feedback above for this 🙂

shawn-ffufm replied:
seems u forgot to remove this 😝

shawn-ffufm commented on lib/utils/color_constants.dart:
```diff
@@ -0,0 +1,7 @@
+import 'package:flutter/material.dart';
+
+const abilityChipIconBackgroundColor = Colors.transparent;
```
kindly just rename the dart file here to just `colors.dart` 🙂

shawn-ffufm commented on lib/utils/utils.dart:
```diff
@@ -0,0 +1,3 @@
+export 'async.dart';
```
same here, see related feedback above for this 🙂

shawn-ffufm commented on lib/widgets/pokemon_profile_card.dart:
```diff
@@ -0,0 +1,34 @@
+    required this.pokemonImage,
+  }) : super(key: key);
+
+  final String pokemonName;
```
just use the whole pokemon object here instead of breaking down its properties

shawn-ffufm commented on lib/widgets/pokemon_profile_card.dart:
```diff
@@ -0,0 +1,34 @@
+  }) : super(key: key);
+
+  final String pokemonName;
+  final String pokemonImage;
```
also, please make it as a habit to put required properties at the top 🙂

shawn-ffufm commented on lib/widgets/pokemon_tile_card.dart:
```diff
@@ -18,16 +19,30 @@ class PokemonTileCard extends StatelessWidget {
+  _navigateToPokemonDetailsPage(BuildContext context, String pokemonImageUrl) {
+    Navigator.of(context).push(
+      MaterialPageRoute(
+        builder: (context) => PokemonDetailsConnector(
```
kindly rename context to _ since it's unused 🙂

shawn-ffufm commented on lib/widgets/widgets.dart:
```diff
@@ -0,0 +1,3 @@
+export 'pokemon_profile_card.dart';
```
same here, see related feedback above for this 🙂

shawn-ffufm reviewed (changes requested):
see pr feedback 🙂

Commit:
feat: implemented pokemon details page
(pr feedback)
- removed added barrel files
- passed on the Pokemon model instead of making it a required property to get pokemon name and url
- modified abilities and moves models where Jsonkey should not be nullable
- renamed class name widgets in pokemon_details folder with a view suffix as well as dart files
- wrapped pokemon details overview in a ListView instead of a SingleChildScrollView
- removed PokemonDetailsOverview class and moved components directly to pokemon_details_page.dart
- separated PhysicalStatsItem widget in a different dart file from pokemon_physical_stats_view.dart
- renamed colors_constants.dart to colors.dart
- renamed context to _ in pokemon_tile_card.dart
- added a reusable string extension for the pokomen image URL path

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_connector.dart:
```diff
@@ -0,0 +1,28 @@
+      onDispose: (store) => store.dispatchSync(ClearPokemonDetailsAction()),
+      builder: (context, vm) => PokemonDetailsPage(
+        pokemonDetails: vm.pokemonDetails,
+        pokemon: pokemon,
```
move this above `pokemonDetails` property 🙂

shawn-ffufm commented on lib/features/pokemon_details/pokemon_details_page.dart:
```diff
@@ -0,0 +1,61 @@
+  });
+
+  final Async<PokemonDetails> pokemonDetails;
+  final Pokemon pokemon;
```
move this above `pokemonDetails` property the same way u did above 🙂

shawn-ffufm commented on lib/features/pokemon_details/widgets/pokemon_physical_stats_view.dart:
```diff
@@ -0,0 +1,41 @@
+    Key? key,
+  }) : super(key: key);
+
+  final PokemonDetails physicalStats;
```
kindly pass only the needed properties instead of passing the whole `PokemonDetails` object, else this is gonna be overkill in a way 😉

shawn-ffufm commented on lib/features/pokemon_details/widgets/pokemon_stats_view.dart:
```diff
@@ -0,0 +1,70 @@
+          physics: const NeverScrollableScrollPhysics(),
+          itemCount: stats.length,
+          itemBuilder: (context, index) {
+            final stat = stats[index].stat;
```
don't do double computation for getting specific stats item 🙈 

would be better if you declare just `final pokemonStats = stats[index];`. Then use it immediately to where it's needed no need to declare separate variables for stat and baseStat unless it's used multiple times 🙂 

just do use it like this directly to where it's needed:
`pokemonStats.stat`
`pokemonStats.baseStat`

shawn-ffufm commented on lib/utils/constants.dart:
```diff
@@ -16,4 +13,13 @@ const pokemonImageUrlPath =
+const String emptyPokemonDetailsLabel = 'No Pokemon Details Available';
 const String emptyString = '';
+
+const String pokemonDetailsTitle = 'Pokemon Details';
```
no need to include its type, same goes for others here 🙂

shawn-ffufm commented on lib/widgets/pokemon_tile_card.dart:
```diff
@@ -14,20 +16,29 @@ class PokemonTileCard extends StatelessWidget {
   Widget build(BuildContext context) {
-    final pokemonId = pokemon.url.split(urlSplitter)[idIndex];
-    final pokemonImageUrl = pokemonImageUrlPath.replaceAll(indexKey, pokemonId);
     final pokemonName = pokemon.name.capitalize();
```
seems no need for this anymore 🙂

jayveeannsoriano replied:
already removed from this commit po 😅

shawn-ffufm reviewed (changes requested):
see pr feedback 🙂

Commit:
feat: implemented pokemon details page
- moved pokemon properties above pokemonDetails
- passed needed properties in pokemon_physical_stats_view instead of the whole object
- removed 2 declarations and changed it to final pokemonStats = stats[index]; in pokemon_stats_view.dart
- removed declared types in constants.dart

Commit:
feat: implemented pokemon details page
(pr feedback)
- removed state barrel file
- added needed imports
